parser.js

```diff
@@ -611,24 +605,36 @@
       return quantifier;
     }
 
-    function parseAtom() {
+    function parseAtomAndExtendedAtom() {
+      // Parsing Atom and ExtendedAtom together due to redundancy.
+      // ExtendedAtom is defined in Apendix B of the ECMA-262 standard.
+      //
+      // SEE: https://www.ecma-international.org/ecma-262/10.0/index.html#prod-annexB-ExtendedPatternCharacter
+      //
       // Atom ::
       //      PatternCharacter
       //      .
       //      \ AtomEscape
       //      CharacterClass
       //      ( GroupSpecifier Disjunction )
       //      ( ? : Disjunction )
+      // ExtendedAtom ::
+      //      ExtendedPatternCharacter
+      // ExtendedPatternCharacter ::
+      //      SourceCharacter but not one of ^$\.*+?()[|
 
       var res;
 
       // jviereck: allow ']', '}' here as well to be compatible with browser's
       //   implementations: ']'.match(/]/);
-      // if (res = matchReg(/^[^^$\\.*+?()[\]{}|]/)) {
-      if (res = matchReg(/^[^^$\\.*+?(){[|]/)) {
+      if (res = matchReg(/^[^^$\\.*+?()[\]{}|]/)) {
         //      PatternCharacter
         return createCharacter(res);
       }
+      else if (!hasUnicodeFlag && (res = matchReg(/^[^^$\\.*+?(){[|]/))) {
```

Comment:
If we are here it means that the character isn't one of `^$\\.*+?()[\]{}|` (because of the prev `if`), and thus it should be enough to check if the character is `]` or `}`.

Reply:
Good point. I was thinking about it initially but then thought the readability of having the full string might be better. Given your comment I just pushed an optimized version as `res = matchReg(/^(?:]|})/))`.